File: internal/pathresolver/pathresolver.go
package pathresolver

import (
	"errors"
	"fmt"
	"strings"
)

var (
	ErrInvalidPath   = errors.New("INVALID_PATH")
	ErrKeyNotFound   = errors.New("KEY_NOT_FOUND")
	ErrPathError     = errors.New("PATH_ERROR")
	ErrPathConflict  = errors.New("PATH_CONFLICT")
	ErrNotObject     = errors.New("NOT_OBJECT")
)

// ValidatePath validates a key path
func ValidatePath(keyPath string) error {
	if keyPath == "" {
		return fmt.Errorf("%w: Key path cannot be empty", ErrInvalidPath)
	}
	return nil
}

// SplitPath splits a dot-notation path into individual keys
func SplitPath(keyPath string) []string {
	if keyPath == "" {
		return []string{}
	}
	return strings.Split(keyPath, ".")
}

// NavigateToKey navigates through nested structure to get value at key path
func NavigateToKey(data interface{}, keyPath string) (interface{}, error) {
	if err := ValidatePath(keyPath); err != nil {
		return nil, err
	}

	// Convert to map[string]interface{} if needed
	dataMap, ok := data.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("%w: Root data is not an object", ErrPathError)
	}

	// First, try the whole path as a single key (handles keys with dots)
	if value, exists := dataMap[keyPath]; exists {
		return value, nil
	}

	// If that fails, try dot-separated navigation
	keys := SplitPath(keyPath)
	current := data

	for i, key := range keys {
		currentMap, ok := current.(map[string]interface{})
		if !ok {
			partialPath := strings.Join(keys[:i], ".")
			return nil, fmt.Errorf("%w: Cannot navigate through non-object value at '%s'", ErrPathError, partialPath)
		}

		value, exists := currentMap[key]
		if !exists {
			return nil, fmt.Errorf("%w: Key '%s' not found", ErrKeyNotFound, keyPath)
		}

		current = value
	}

	return current, nil
}

// NavigateToParent navigates to the parent of the target key
func NavigateToParent(data interface{}, keyPath string) (map[string]interface{}, string, error) {
	if err := ValidatePath(keyPath); err != nil {
		return nil, "", err
	}

	keys := SplitPath(keyPath)
	
	dataMap, ok := data.(map[string]interface{})
	if !ok {
		return nil, "", fmt.Errorf("%w: Root data is not an object", ErrPathError)
	}

	if len(keys) == 1 {
		// Key is at root level
		return dataMap, keys[0], nil
	}

	// Navigate to parent
	parentPath := strings.Join(keys[:len(keys)-1], ".")
	parent, err := NavigateToKey(data, parentPath)
	if err != nil {
		return nil, "", err
	}

	parentMap, ok := parent.(map[string]interface{})
	if !ok {
		return nil, "", fmt.Errorf("%w: Parent at '%s' is not an object", ErrPathError, parentPath)
	}

	return parentMap, keys[len(keys)-1], nil
}

// KeyExists checks if a key path exists in the data structure
func KeyExists(data interface{}, keyPath string) bool {
	if err := ValidatePath(keyPath); err != nil {
		return false
	}

	dataMap, ok := data.(map[string]interface{})
	if !ok {
		return false
	}

	// First, try the whole path as a single key (handles keys with dots)
	if _, exists := dataMap[keyPath]; exists {
		return true
	}

	// If that fails, try dot-separated navigation
	_, err := NavigateToKey(data, keyPath)
	return err == nil
}

// CreateNestedPath creates nested path structure, creating intermediate objects as needed
func CreateNestedPath(data map[string]interface{}, keyPath string) (map[string]interface{}, error) {
	if err := ValidatePath(keyPath); err != nil {
		return nil, err
	}

	keys := SplitPath(keyPath)
	
	if len(keys) == 1 {
		// Root level key
		return data, nil
	}

	current := data

	// Create intermediate objects
	for i, key := range keys[:len(keys)-1] {
		if value, exists := current[key]; exists {
			if valueMap, ok := value.(map[string]interface{}); ok {
				current = valueMap
			} else {
				partialPath := strings.Join(keys[:i+1], ".")
				return nil, fmt.Errorf("%w: Cannot create nested path through non-object value at '%s'", ErrPathConflict, partialPath)
			}
		} else {
			// Create new nested object
			newMap := make(map[string]interface{})
			current[key] = newMap
			current = newMap
		}
	}

	return current, nil
}

// GetAllKeysAtPath gets all immediate child keys at the specified path
func GetAllKeysAtPath(data interface{}, keyPath *string) ([]string, error) {
	var target interface{}
	var err error

	if keyPath == nil {
		target = data
	} else {
		target, err = NavigateToKey(data, *keyPath)
		if err != nil {
			return nil, err
		}
	}

	targetMap, ok := target.(map[string]interface{})
	if !ok {
		pathDesc := "root"
		if keyPath != nil {
			pathDesc = fmt.Sprintf("'%s'", *keyPath)
		}
		return nil, fmt.Errorf("%w: Value at %s is not an object, cannot list keys", ErrNotObject, pathDesc)
	}

	keys := make([]string, 0, len(targetMap))
	for key := range targetMap {
		keys = append(keys, key)
	}

	return keys, nil
}

// RemoveKeyAtPath removes a key at the specified path and returns its value
func RemoveKeyAtPath(data map[string]interface{}, keyPath string) (interface{}, error) {
	if err := ValidatePath(keyPath); err != nil {
		return nil, err
	}

	// First, try the whole path as a single key (handles keys with dots)
	if value, exists := data[keyPath]; exists {
		delete(data, keyPath)
		return value, nil
	}

	// If that fails, try dot-separated navigation
	parent, finalKey, err := NavigateToParent(data, keyPath)
	if err != nil {
		return nil, err
	}

	value, exists := parent[finalKey]
	if !exists {
		return nil, fmt.Errorf("%w: Key '%s' not found", ErrKeyNotFound, keyPath)
	}

	delete(parent, finalKey)
	return value, nil
}

// SetValueAtPath sets a value at the specified path
func SetValueAtPath(data map[string]interface{}, keyPath string, value interface{}, createPath bool) error {
	var parent map[string]interface{}
	var finalKey string
	var err error

	if createPath {
		parent, err = CreateNestedPath(data, keyPath)
		if err != nil {
			return err
		}
		finalKey = SplitPath(keyPath)[len(SplitPath(keyPath))-1]
	} else {
		parent, finalKey, err = NavigateToParent(data, keyPath)
		if err != nil {
			return err
		}
	}

	parent[finalKey] = value
	return nil
}

File: src/path_resolver.py
"""
Path Resolver Module

Handles dot-notation key path resolution for nested JSON structures.
Supports complex path navigation and validation.
"""

from typing import Any, Dict, List, Optional, Union


class PathResolver:
    """Handles path resolution for nested JSON structures"""
    
    @staticmethod
    def validate_path(key_path: Optional[str]) -> str:
        """
        Validate a key path
        
        Args:
            key_path: The key path to validate
            
        Returns:
            Validated key path (unchanged)
            
        Raises:
            ValueError: If path is invalid
        """
        if key_path is None:
            raise ValueError("INVALID_PATH: Key path cannot be None")
        
        if not isinstance(key_path, str):
            raise ValueError("INVALID_PATH: Key path must be a string")
        
        if key_path == '':
            raise ValueError("INVALID_PATH: Key path cannot be empty")
        
        return key_path
    
    @staticmethod
    def split_path(key_path: str) -> List[str]:
        """
        Split a dot-notation path into individual keys
        
        Args:
            key_path: Dot-notation path (e.g., 'section.subsection.key')
            
        Returns:
            List of individual keys
            
        Examples:
            split_path('a.b.c') -> ['a', 'b', 'c']
            split_path('simple') -> ['simple']
        """
        if not key_path:
            return []
        
        return key_path.split('.')
    
    @staticmethod
    def navigate_to_key(data: Dict[str, Any], key_path: str) -> Any:
        """
        Navigate through nested structure to get value at key path
        
        Args:
            data: Root JSON data dictionary
            key_path: Dot-notation path to navigate
            
        Returns:
            Value at the specified path
            
        Raises:
            KeyError: If key path doesn't exist
            TypeError: If trying to navigate through non-dict value
        """
        validated_path = PathResolver.validate_path(key_path)
        
        # First, try the whole path as a single key (handles keys with dots)
        if isinstance(data, dict) and validated_path in data:
            return data[validated_path]
        
        # If that fails, try dot-separated navigation
        keys = PathResolver.split_path(validated_path)
        current = data
        
        for i, key in enumerate(keys):
            if not isinstance(current, dict):
                partial_path = '.'.join(keys[:i])
                raise TypeError(f"PATH_ERROR: Cannot navigate through non-object value at '{partial_path}'")
            
            if key not in current:
                raise KeyError(f"KEY_NOT_FOUND: Key '{key_path}' not found")
            
            current = current[key]
        
        return current
    
    @staticmethod
    def navigate_to_parent(data: Dict[str, Any], key_path: str) -> tuple[Dict[str, Any], str]:
        """
        Navigate to the parent of the target key
        
        Args:
            data: Root JSON data dictionary
            key_path: Dot-notation path to target key
            
        Returns:
            Tuple of (parent_dict, final_key)
            
        Raises:
            KeyError: If parent path doesn't exist
            TypeError: If trying to navigate through non-dict value
        """
        validated_path = PathResolver.validate_path(key_path)
        keys = PathResolver.split_path(validated_path)
        
        if len(keys) == 1:
            # Key is at root level
            return data, keys[0]
        
        # Navigate to parent
        parent_path = '.'.join(keys[:-1])
        parent = PathResolver.navigate_to_key(data, parent_path)
        
        if not isinstance(parent, dict):
            raise TypeError(f"PATH_ERROR: Parent at '{parent_path}' is not an object")
        
        return parent, keys[-1]
    
    @staticmethod
    def key_exists(data: Dict[str, Any], key_path: str) -> bool:
        """
        Check if a key path exists in the data structure
        
        Args:
            data: Root JSON data dictionary
            key_path: Dot-notation path to check
            
        Returns:
            True if key exists, False otherwise
        """
        try:
            validated_path = PathResolver.validate_path(key_path)
            
            # First, try the whole path as a single key (handles keys with dots)
            if isinstance(data, dict) and validated_path in data:
                return True
            
            # If that fails, try dot-separated navigation
            PathResolver.navigate_to_key(data, key_path)
            return True
        except (KeyError, TypeError, ValueError):
            return False
    
    @staticmethod
    def create_nested_path(data: Dict[str, Any], key_path: str) -> Dict[str, Any]:
        """
        Create nested path structure, creating intermediate objects as needed
        
        Args:
            data: Root JSON data dictionary (modified in place)
            key_path: Dot-notation path to create
            
        Returns:
            The parent dictionary where the final key should be set
            
        Raises:
            ValueError: If path conflicts with existing non-object value
        """
        validated_path = PathResolver.validate_path(key_path)
        keys = PathResolver.split_path(validated_path)
        
        if len(keys) == 1:
            # Root level key
            return data
        
        current = data
        
        # Create intermediate objects
        for i, key in enumerate(keys[:-1]):
            if key not in current:
                current[key] = {}
            elif not isinstance(current[key], dict):
                partial_path = '.'.join(keys[:i+1])
                raise ValueError(f"PATH_CONFLICT: Cannot create nested path through non-object value at '{partial_path}'")
            
            current = current[key]
        
        return current
    
    @staticmethod
    def get_all_keys_at_path(data: Dict[str, Any], key_path: Optional[str] = None) -> List[str]:
        """
        Get all immediate child keys at the specified path
        
        Args:
            data: Root JSON data dictionary
            key_path: Path to get keys from (None for root level)
            
        Returns:
            List of immediate child key names
            
        Raises:
            KeyError: If path doesn't exist
            TypeError: If path points to non-object value
        """
        if key_path is None:
            target = data
        else:
            target = PathResolver.navigate_to_key(data, key_path)
        
        if not isinstance(target, dict):
            path_desc = "root" if key_path is None else f"'{key_path}'"
            raise TypeError(f"NOT_OBJECT: Value at {path_desc} is not an object, cannot list keys")
        
        return list(target.keys())
    
    @staticmethod
    def remove_key_at_path(data: Dict[str, Any], key_path: str) -> Any:
        """
        Remove a key at the specified path and return its value
        
        Args:
            data: Root JSON data dictionary (modified in place)
            key_path: Path to key to remove
            
        Returns:
            The removed value
            
        Raises:
            KeyError: If key doesn't exist
        """
        validated_path = PathResolver.validate_path(key_path)
        
        # First, try the whole path as a single key (handles keys with dots)
        if isinstance(data, dict) and validated_path in data:
            return data.pop(validated_path)
        
        # If that fails, try dot-separated navigation
        parent, final_key = PathResolver.navigate_to_parent(data, key_path)
        
        if final_key not in parent:
            raise KeyError(f"KEY_NOT_FOUND: Key '{key_path}' not found")
        
        return parent.pop(final_key)
    
    @staticmethod
    def set_value_at_path(data: Dict[str, Any], key_path: str, value: Any, create_path: bool = False) -> None:
        """
        Set a value at the specified path
        
        Args:
            data: Root JSON data dictionary (modified in place)
            key_path: Path where to set the value
            value: Value to set
            create_path: Whether to create intermediate objects if they don't exist
            
        Raises:
            KeyError: If path doesn't exist and create_path is False
            ValueError: If path conflicts with existing structure
        """
        if create_path:
            parent = PathResolver.create_nested_path(data, key_path)
            final_key = PathResolver.split_path(key_path)[-1]
        else:
            parent, final_key = PathResolver.navigate_to_parent(data, key_path)
        
        parent[final_key] = value